recipe - CategoricalAnomalyDetector with loda

KrisCowie opened the pull request.

woozyking commented:
this looks good! a few suggestions:

* try to leverage the [recent development](https://github.com/EQWorks/locussdk/pull/138) on unified GeoJSON retrieval to obtain lightweight (but often sufficient) polygon shapes for FSA/postal codes
* I would definitely encourage you to retain the change in the `recipes/categorical_anomaly.py` module directly to reflect the support of the LODA model (this way we can use the PR review tool to examine the changeset more in detail)
* we can probably merge the LODA model usage demo in the same context as the previous isolation forest one along with the first suggestion. also perhaps also show the map-based visualization (to help spot out anomaly in a geo-context) along with the non-geo version

KrisCowie commented:
> this looks good! a few suggestions:
> 
> * try to leverage the [recent development](https://github.com/EQWorks/locussdk/pull/138) on unified GeoJSON retrieval to obtain lightweight (but often sufficient) polygon shapes for FSA/postal codes
> * I would definitely encourage you to retain the change in the `recipes/categorical_anomaly.py` module directly to reflect the support of the LODA model (this way we can use the PR review tool to examine the changeset more in detail)
> * we can probably merge the LODA model usage demo in the same context as the previous isolation forest one along with the first suggestion. also perhaps also show the map-based visualization (to help spot out anomaly in a geo-context) along with the non-geo version

Hey @woozyking , so I've added Local Outlier Factor, and OneClass SVM as options, as well as the option to autotune for LOF, and finally all four - Isolation Forest, LODA, SVM, and LOF - have the option to use multiple metrics column instead of just one 

They can all use the choropleth mapping as well.

I didn't add the get GeoJson, wasn't really sure where that could be used in the class? Unless you meant just in the notebook use that instead of the shapefile I found? 
I added the datasets just into a folder so you can run it and check functionality, we can delete those whenever.

Oh and I couldn't quite get the file small enough with the choropleth maps actually loaded, so you'd have to run them locally - if you run the cells in sequence they'll load the choropleth map, just uploading to git was too big.

woozyking commented:
> I didn't add the get GeoJson, wasn't really sure where that could be used in the class? Unless you meant just in the notebook use that instead of the shapefile I found?

yea, in the notebook, try to use the locussdk function to query for the interested FSAs or postal codes (you may need to show the way to work around per-query limit of the number of POIs)

this would also reduce the size of the choropleth visualization because the polygon shapes from locussdk function has a much more optimal size

KrisCowie commented:
> yea, in the notebook, try to use the locussdk function to query for the interested FSAs or postal codes (you may need to show the way to work around per-query limit of the number of POIs)
> 
> this would also reduce the size of the choropleth visualization because the polygon shapes from locussdk function has a much more optimal size

Ooooh gotcha! Haha ya I was like huh, not sure how to slot that into this class, then as I was uploading went wait, maybe he meant just in the notebook. 
That makes sense haha. The file was 450 mb when I tried to upload with 5 maps and the scatter plots, don't think it'll be able to have much more than one or two examples with the maps even with the smaller size? I'll check it out, unos momentos

woozyking commented:
noice. once you test it out, perhaps remove the source data (including the shapefiles and the data csv files), as this is mean to demonstrate general capabilities (so ppl can plug in their own data to try out the core classes)

KrisCowie commented:
> noice. once you test it out, perhaps remove the source data (including the shapefiles and the data csv files), as this is mean to demonstrate general capabilities (so ppl can plug in their own data to try out the core classes)

Hokie dokie, all done! I included an ontario.geojson, but could change that to just a list of FSAs if you want? It was just to trim the initial list of Canadian FSA's to Ontario, get_geojson is still getting the actual polygons.
Other than that, it's all good to go. I tested as many configs as I could think of - all four algos, LOF with and without automatic tuning, all with one metric, with multiple metrics, all with choropleths maps, all with scatter plots, and finally with uniform manifold approximation. Let me know what ya think :)

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -68,18 +223,25 @@ def graph(self, no_geo: bool = False, **kwargs):
+                'height': 720,
+                **kwargs,
+            }
+            return px.choropleth_mapbox(**plotly_args)
```
is there an else case to cover outside of `if self.model_type:`?

also please ensure an empty new line per unix conventions. you can usually configure your editor to do that automatically on-save.

KrisCowie replied:
Ah shoot good point, I'll toss something in like "Please select an appropriate model type".
Oh sorry, that's weird, didn't know macs did that, the other repos must've had gitignore .DS_Store already. 
Kk, added and deleted.

KrisCowie replied:
~~It didn't include the space at the end. 
I have it on my local, and it's definitely the version I just pushed! What the heck?~~

Never mind it just wasn't showing in the review, it's there in the actual file

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+            self.model = OneClassSVM(**model_args)
+            return self.model
+        if self.model_type == 'LocalOutlierFactor':
+            res = input("Would you like to automatically tune hyperparameters? ")
```
use parameters instead of interactive prompts. this way, one can use this method fully programmatically, and if they wish to add the interactivity, they _can_ do so in their own app layer:

```python
# method signature
def create_model(self, auto_tune: bool = True, **model_args):
    ....
```

```python
# imaginary application using this class
from ... import CategoricalAnomalyDetector

detector = CategoricalAnomalyDetector(...)
...
# interactivity
res = input("Would you like to automatically tune hyperparameters? ")
auto_tune = res == 'yes'  # logic to produce the bool signal

# the method abstractly cares only about the definitive bool signal
detector.create_model(auto_tune=auto_tune)
```

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+        if self.model_type == 'LocalOutlierFactor':
+            res = input("Would you like to automatically tune hyperparameters? ")
+            if res == 'yes':
+                print("Automatically tuning parameters")
```
do not over-print. if needed add a `debug` parameter (or `_debug` for the undocumented parameter) and `logging` library to log debug messages when opt-in

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+        if self.model_type == 'IsolationForest':
+            self.model = IsolationForest(**model_args)
+            return self.model
+
```
the `return` can be unified at the end of the method, as it only returns a object/state variable mutated by the `if` conditions:

```python
if not self.model_type:
    raise Exception(...)

if self.model_type == '...':
    self.model = ...
elif self.model_type == '...':
    ...
elif ...:
    ...

return self.model
```

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+            score_df['score'] = self.model.fit_predict(score_df[[self.metrics_column]].values)
+            score_df['anomaly'] = score_df['score'] <= -1
+            self.score_df = score_df
+            return score_df
```
similar as the comment for the `create_model()` method, both mutation of object/state variable `self.score_df` and `return` can be unified at the end, and the df copy action can happen in the beginning:

```python
score_df = df.copy()
if ...
    ...things that mutate local score_df
elif ...
    ....
else
    ....

self.score_df = score_df
return score_df
```

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+        if (three_d and no_geo and ((self.model_type == 'LODA') or type(self.metrics_column) == list)):
+            if self.model_type == 'LODA':
+                numeric_features = list(self.df.select_dtypes(include=['int64', 'float64']).columns)
+                X = self.score_df[numeric_features].values
```
try to stick to the [PEP8 conventions](https://www.python.org/dev/peps/pep-0008/), name variables in `lower_case_with_underscores` and try to give it either a slightly more elaborate name to be self-documenting/explanatory, or provide inline documentation to reference what `x` means

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+                title=f'Total Explained Variance: {total_var:.2f}%',
+                hover_data = [self.category_column],
+                )
+            return fig.show()
```
similar comment, as `fig` is being manipulated throughout the method, its invocation can be unified toward the end

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
             return self.score_df.loc[self.score_df['anomaly'] == True]
 
-    def graph(self, no_geo: bool = False, **kwargs):
+    def graph(self, no_geo: bool = False, three_d: bool = False, **kwargs):
```
as the method complexity grow, we can potentially split it into 3 distinct ones:

```python
def graph_geo(...)

def graph_3d(...)

def graph(...)
```

this way, each tends to be a bit simpler in scope, thus easier to reason-with and maintain, as there are less flow control logic which tends to lead to overlooked bugs (exhibit A, the previously missing `else` case for some other method)

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -33,30 +47,219 @@ def scale(self, ScalerClass = None):
+                )
+            return fig.show()
+        
+        if (no_geo and ((self.model_type == 'LODA') or type(self.metrics_column) == list)):
```
`type(...)` output is often compared using `is` (similar to `===` in some other languages):

```python
or type(self.metrics_column) is list)
```

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -68,18 +271,27 @@ def graph(self, no_geo: bool = False, **kwargs):
+            }
+            return px.choropleth_mapbox(**plotly_args)
+        else:
+            return("Please input an appropriate model name")
```
`return` serves as an early exit, thus, there is no need for the `else`

however, the lack of `model_type` seems that it should be treated more like an exception, and can be raised so (and `raise` also serves as an early exit).

as python uses indentation to enforce flow control, good utilization of "early exit" technique is thus even more important. so combining together with ^

```python
# early exit on a smaller scope of logic
if not self.model_type:
    raise Exception('Missing model type')

# retain one indentation level for the larger scope of logic
...
plotly_args = ...
return px.chropleth_mapbox(**plotly_args)
```

woozyking commented:
also changing the base to `main` as the og anomaly detection branch has been merged already

woozyking commented on recipes/categorical_anomaly.py:
```diff
@@ -124,143 +124,139 @@ def tune_params(self,df, metrics_column):
 
 
-    def create_model(self, **model_args):
+    def create_model(self,auto_tune=None, **model_args):
```
make sure to have a space after `,`, and around `=` to conform to the PEP8 conventions:

```python
def create_model(self, auto_tune: bool = None, **model_args):
```

woozyking replied:
you can probably use flake8 locally to flag those PEP8 violations automatically, see some other Python repos under the org for reference

KrisCowie replied:
Ah kk will do! Haha and also remember to do a once over before I upload, I tend to get excited when everything works after all the permutation testing, but forget to double check the little stuff that makes the code readable and normalized so others can read it without their eyes bleeding.
Haha basically just slow down, have a PEP8 cheat sheet up on top of flake8, and take it easy :)

woozyking replied:
ye you can also install https://pypi.org/project/autopep8/ locally which automatically addresses some of the issues
